File: apps/brittney-desktop/src-tauri/src/inference/ollama.rs
//! Ollama Client for Local LLM Inference
//!
//! Connects to local Ollama server (default: localhost:11434)
//! Provides chat completion with Brittney models

use super::{ChatMessage, InferenceRequest, InferenceResponse, ProviderStatus, Usage};
use anyhow::{Result, anyhow};
use reqwest::Client;
use serde::{Deserialize, Serialize};

const DEFAULT_OLLAMA_URL: &str = "http://localhost:11434";

#[derive(Debug, Clone)]
pub struct OllamaClient {
    client: Client,
    base_url: String,
    timeout_secs: u64,
}

#[derive(Debug, Serialize)]
struct OllamaChatRequest {
    model: String,
    messages: Vec<OllamaMessage>,
    stream: bool,
    options: OllamaOptions,
}

#[derive(Debug, Serialize)]
struct OllamaMessage {
    role: String,
    content: String,
}

#[derive(Debug, Serialize)]
struct OllamaOptions {
    temperature: f32,
    num_predict: u32,
}

#[derive(Debug, Deserialize)]
struct OllamaChatResponse {
    model: String,
    message: OllamaResponseMessage,
    done: bool,
    #[serde(default)]
    prompt_eval_count: u32,
    #[serde(default)]
    eval_count: u32,
}

#[derive(Debug, Deserialize)]
struct OllamaResponseMessage {
    role: String,
    content: String,
}

#[derive(Debug, Deserialize)]
struct OllamaModelsResponse {
    models: Vec<OllamaModel>,
}

#[derive(Debug, Deserialize)]
struct OllamaModel {
    name: String,
    #[serde(default)]
    size: u64,
}

impl OllamaClient {
    pub fn new(base_url: Option<String>, timeout_secs: Option<u64>) -> Self {
        Self {
            client: Client::new(),
            base_url: base_url.unwrap_or_else(|| DEFAULT_OLLAMA_URL.to_string()),
            timeout_secs: timeout_secs.unwrap_or(120),
        }
    }

    /// Check if Ollama is running and available
    pub async fn health(&self) -> bool {
        let url = format!("{}/api/tags", self.base_url);
        match self.client.get(&url)
            .timeout(std::time::Duration::from_secs(5))
            .send()
            .await
        {
            Ok(resp) => resp.status().is_success(),
            Err(_) => false,
        }
    }

    /// Get status with available models
    pub async fn get_status(&self) -> ProviderStatus {
        let url = format!("{}/api/tags", self.base_url);

        match self.client.get(&url)
            .timeout(std::time::Duration::from_secs(5))
            .send()
            .await
        {
            Ok(resp) if resp.status().is_success() => {
                match resp.json::<OllamaModelsResponse>().await {
                    Ok(data) => ProviderStatus {
                        provider: "local".to_string(),
                        available: true,
                        error: None,
                        models: data.models.into_iter().map(|m| m.name).collect(),
                    },
                    Err(e) => ProviderStatus {
                        provider: "local".to_string(),
                        available: false,
                        error: Some(format!("Failed to parse models: {}", e)),
                        models: vec![],
                    },
                }
            },
            Ok(resp) => ProviderStatus {
                provider: "local".to_string(),
                available: false,
                error: Some(format!("Ollama returned status: {}", resp.status())),
                models: vec![],
            },
            Err(e) => ProviderStatus {
                provider: "local".to_string(),
                available: false,
                error: Some(format!("Ollama not running: {}", e)),
                models: vec![],
            },
        }
    }

    /// Check if a specific model is available
    pub async fn has_model(&self, model_name: &str) -> bool {
        let status = self.get_status().await;
        status.models.iter().any(|m| m.starts_with(model_name.split(':').next().unwrap_or(model_name)))
    }

    /// Chat completion
    pub async fn chat(&self, request: InferenceRequest) -> Result<InferenceResponse> {
        let url = format!("{}/api/chat", self.base_url);

        let model = request.model.unwrap_or_else(|| "brittney-v4-expert:latest".to_string());

        let ollama_request = OllamaChatRequest {
            model: model.clone(),
            messages: request.messages.into_iter().map(|m| OllamaMessage {
                role: m.role,
                content: m.content,
            }).collect(),
            stream: false,
            options: OllamaOptions {
                temperature: request.temperature,
                num_predict: request.max_tokens,
            },
        };

        let resp = self.client.post(&url)
            .timeout(std::time::Duration::from_secs(self.timeout_secs))
            .json(&ollama_request)
            .send()
            .await?;

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().await.unwrap_or_default();
            return Err(anyhow!("Ollama error {}: {}", status, text));
        }

        let ollama_resp: OllamaChatResponse = resp.json().await?;

        Ok(InferenceResponse {
            id: format!("ollama_{}", chrono::Utc::now().timestamp_millis()),
            content: ollama_resp.message.content,
            model,
            provider: "local".to_string(),
            usage: Some(Usage {
                prompt_tokens: ollama_resp.prompt_eval_count,
                completion_tokens: ollama_resp.eval_count,
                total_tokens: ollama_resp.prompt_eval_count + ollama_resp.eval_count,
            }),
        })
    }

    /// Pull a model from Ollama registry
    pub async fn pull_model(&self, model_name: &str) -> Result<()> {
        let url = format!("{}/api/pull", self.base_url);

        let resp = self.client.post(&url)
            .json(&serde_json::json!({ "name": model_name }))
            .send()
            .await?;

        if !resp.status().is_success() {
            return Err(anyhow!("Failed to pull model: {}", resp.status()));
        }

        // Read the streaming response (just consume it for now)
        let _ = resp.text().await?;

        Ok(())
    }
}

impl Default for OllamaClient {
    fn default() -> Self {
        Self::new(None, None)
    }
}

File: packages/shared/inference/src/providers/ollama.ts
/**
 * Ollama Provider - Local LLM Inference
 *
 * Connects to Ollama server (default: localhost:11434)
 * Used by both Hololand app and MCP tools
 */

import type {
  ChatMessage,
  InferenceRequest,
  InferenceResponse,
  StreamChunk,
  ModelInfo,
  ProviderStatus,
  ProviderType,
} from '../types.js';

// API Response types
interface OllamaTagsResponse {
  models?: OllamaModel[];
}

interface OllamaChatResponse {
  message?: { content: string };
  response?: string;
  prompt_eval_count?: number;
  eval_count?: number;
}

interface OllamaStreamChunk {
  message?: { content: string };
  done?: boolean;
}

export interface OllamaConfig {
  baseUrl: string;
  timeout: number;
}

export interface OllamaModel {
  name: string;
  modified_at: string;
  size: number;
  digest: string;
  details?: {
    format?: string;
    family?: string;
    parameter_size?: string;
    quantization_level?: string;
  };
}

export class OllamaProvider {
  private baseUrl: string;
  private timeout: number;

  constructor(config: Partial<OllamaConfig> = {}) {
    this.baseUrl = config.baseUrl || 'http://localhost:11434';
    this.timeout = config.timeout || 120000;
  }

  get type(): ProviderType {
    return 'local';
  }

  /**
   * Check if Ollama is available
   */
  async health(): Promise<boolean> {
    try {
      const response = await fetch(`${this.baseUrl}/api/tags`, {
        signal: AbortSignal.timeout(5000),
      });
      return response.ok;
    } catch {
      return false;
    }
  }

  /**
   * Get provider status with available models
   */
  async getStatus(): Promise<ProviderStatus> {
    const start = Date.now();
    try {
      const available = await this.health();
      if (!available) {
        return { type: 'local', available: false, error: 'Ollama not running' };
      }

      const models = await this.listModels();
      return {
        type: 'local',
        available: true,
        models: models.map((m) => this.toModelInfo(m)),
        latencyMs: Date.now() - start,
      };
    } catch (error: any) {
      return { type: 'local', available: false, error: error.message };
    }
  }

  /**
   * List available models
   */
  async listModels(): Promise<OllamaModel[]> {
    const response = await fetch(`${this.baseUrl}/api/tags`);
    if (!response.ok) {
      throw new Error(`Failed to list models: ${response.status}`);
    }
    const data = (await response.json()) as OllamaTagsResponse;
    return data.models || [];
  }

  /**
   * Check if a specific model is available
   */
  async hasModel(modelName: string): Promise<boolean> {
    const models = await this.listModels();
    return models.some((m) => m.name === modelName || m.name.startsWith(modelName.split(':')[0]));
  }

  /**
   * Pull a model from Ollama registry
   */
  async pullModel(modelName: string, onProgress?: (status: string) => void): Promise<void> {
    const response = await fetch(`${this.baseUrl}/api/pull`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ name: modelName }),
    });

    if (!response.ok) {
      throw new Error(`Failed to pull model: ${response.status}`);
    }

    const reader = response.body?.getReader();
    if (!reader) return;

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      const text = new TextDecoder().decode(value);
      const lines = text.split('\n').filter((l) => l.trim());

      for (const line of lines) {
        try {
          const data = JSON.parse(line);
          if (data.status && onProgress) {
            onProgress(data.status);
          }
        } catch {
          // Ignore parse errors
        }
      }
    }
  }

  /**
   * Chat completion
   */
  async chat(request: InferenceRequest): Promise<InferenceResponse> {
    const start = Date.now();
    const model = request.model || 'brittney-qwen-v23:latest';

    const requestBody = {
      model,
      messages: request.messages,
      options: {
        temperature: request.temperature ?? 0.7,
        num_predict: request.maxTokens ?? 2048,
      },
      stream: false,
    };

    console.log(`[Ollama] Chat request:`, {
      url: `${this.baseUrl}/api/chat`,
      model,
      messageCount: request.messages.length
    });

    const response = await fetch(`${this.baseUrl}/api/chat`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(requestBody),
      signal: AbortSignal.timeout(this.timeout),
    });

    if (!response.ok) {
      const errorBody = await response.text().catch(() => 'Unable to read error response');
      console.error(`[Ollama] Chat failed:`, {
        status: response.status,
        statusText: response.statusText,
        url: `${this.baseUrl}/api/chat`,
        model,
        errorBody
      });
      throw new Error(`Ollama chat failed: ${response.status} ${response.statusText}`);
    }

    const data = (await response.json()) as OllamaChatResponse;

    return {
      id: `ollama_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`,
      content: data.message?.content || data.response || '',
      model,
      provider: 'local',
      usage: {
        promptTokens: data.prompt_eval_count || 0,
        completionTokens: data.eval_count || 0,
        totalTokens: (data.prompt_eval_count || 0) + (data.eval_count || 0),
      },
      latencyMs: Date.now() - start,
    };
  }

  /**
   * Streaming chat completion
   */
  async *chatStream(request: InferenceRequest): AsyncGenerator<StreamChunk> {
    const model = request.model || 'brittney-qwen-v23:latest';

    const response = await fetch(`${this.baseUrl}/api/chat`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        model,
        messages: request.messages,
        options: {
          temperature: request.temperature ?? 0.7,
          num_predict: request.maxTokens ?? 2048,
        },
        stream: true,
      }),
      signal: AbortSignal.timeout(this.timeout),
    });

    if (!response.ok) {
      throw new Error(`Ollama stream failed: ${response.status}`);
    }

    const reader = response.body?.getReader();
    if (!reader) {
      yield { content: '', done: true };
      return;
    }

    const decoder = new TextDecoder();

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      const text = decoder.decode(value);
      const lines = text.split('\n').filter((l) => l.trim());

      for (const line of lines) {
        try {
          const data = JSON.parse(line);
          if (data.message?.content) {
            yield { content: data.message.content, done: false };
          }
          if (data.done) {
            yield { content: '', done: true };
            return;
          }
        } catch {
          // Ignore parse errors
        }
      }
    }

    yield { content: '', done: true };
  }

  /**
   * Generate embeddings
   */
  async embeddings(text: string, model: string = 'nomic-embed-text'): Promise<number[]> {
    const response = await fetch(`${this.baseUrl}/api/embeddings`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ model, prompt: text }),
    });

    if (!response.ok) {
      throw new Error(`Embeddings failed: ${response.status}`);
    }

    const data = (await response.json()) as { embedding: number[] };
    return data.embedding;
  }

  /**
   * Convert Ollama model to ModelInfo
   */
  private toModelInfo(model: OllamaModel): ModelInfo {
    const isBrittney = model.name.toLowerCase().includes('brittney');

    return {
      name: model.name,
      displayName: model.name.split(':')[0],
      provider: 'local',
      size: model.size,
      quantization: model.details?.quantization_level,
      capabilities: isBrittney ? ['chat', 'code', 'holoscript'] : ['chat', 'code'],
      isLocal: true,
      isDownloaded: true,
    };
  }
}
